### PR TITLE
More refactoring of MinHash comparison code

### LINKED ISSUE #868: refactor compare and similarity in _minhash.py and sketch/minhash.rs
Continuing on the mission of #856, these functions could use some cleanup.

A few thoughts --
* combine if/else statements per https://github.com/dib-lab/sourmash/pull/856#discussion_r371029486
* factor out the similarity metrics to Jaccard and angular similarity functions

### PR DESCRIPTION
**Pull request by ctb**
Fixes #868.

* create rust functions `jaccard` and `angular_similarity` to do the actual math on compatible signatures
* make Python `compare` and `similarity` call the same `compare` function on rust side
* refactor out duplicated code for signature downsampling to new rust function `downsample_max_hash`

I made the decision to just simplify and rationalize the rust side of things, because our Python API for MinHash objects (in `sourmash/_minhash.py`) probably shouldn't change without a major version release...

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
Ready for review @luizirber, and thx for the suggestion, @kloetzl !

**Comment by kloetzl** [comment]
Uh, nice. That really reduced the line count. Less lines, less bugs!

**Comment by ctb** [comment]
On Sun, Feb 02, 2020 at 05:36:17AM -0800, Fabian Klötzl wrote:
> Uh, nice. That really reduced the line count. Less lines, less bugs!

yep, thank you!

File: src/core/src/sketch/minhash.rs [hunk 1]
```diff
@@ -460,30 +460,12 @@ impl KmerMinHash {
...
-                    self.seed,
-                    other.max_hash,
-                    self.abunds.is_some(),
-                );
-                new_mh.add_many(&self.mins)?;
-                new_mh.count_common(other, false)
-            }
+            let cmp = self.max_hash < other.max_hash;
+            let a = if cmp { self } else { other };
+            let b = if cmp { other } else { self };
```
Comment (luizirber) [inline src/core/src/sketch/minhash.rs]: I think this is better, but still a bit confusing... How about
```rust
let (a, b) = if self.max_hash < other.max_hash {                                                                                                                                
  (self, other)                                                                                                                                                               
} else {                                                                                                                                                                        
  (other, self)                                                                                                                                                               
};  
```
(`a` and `b` are also not very good variable names, but not sure what to suggest =])

File: src/core/src/sketch/minhash.rs [hunk 2]
```diff
@@ -543,128 +525,95 @@ impl KmerMinHash {
...
+        self.check_compatible(other)?;
+        if let Ok((common, size)) = self.intersection_size(other) {
+            Ok(common as f64 / u64::max(1, size) as f64)
+        } else {
+            Ok(0.0)
+        }
+    }
     // compare two minhashes, ignoring abundance.
+    // CTB: this is unused by ffi now.
+    // CTB: this can be replaced by a call to similarity(..., ignore_abundance=True)
```
Comment (luizirber) [inline src/core/src/sketch/minhash.rs]: I'm OK with removing this on the Rust side, and update tests. Also need to bump version to `0.5.0` in `src/core/Cargo.toml` =]

File: src/core/src/sketch/minhash.rs (hunk 1 above)
Comment (ctb) [reply src/core/src/sketch/minhash.rs]: oooh tuple unpacking!!

File: src/core/src/sketch/minhash.rs (hunk 2 above)
Comment (ctb) [reply src/core/src/sketch/minhash.rs]: done in ed6f43a1c

**Comment by ctb** [comment]
OK I _think_ this is ready.

**Comment by luizirber** [comment]
I'll fix the `wasm-pack` issue. Should I also fix all the failing Rust tests?

**Pull request merged**